### PR TITLE
Fix/ebrake audit fixes

### PR DESCRIPTION
**Pull request by alcueca**
Following the recommendation to use a different Plan struct led to a complete rewrite of the contract, made more complex because apparently you can't add dynamic arrays inside mappings. You can nest mappings, so I made my own dynamic array 🤷‍♂️

### CODE REVIEW

**Comment by alcueca** [comment]
Still have to fix the tests, obviously.

**Comment by alcueca** [comment]
Good to go!

**Comment by alcueca** [comment]
Added a couple of features after developing the governance proposals to enable it.

**Comment by alcueca** [comment]
Fixed the issue around index. Not sure what to do with `permissionAt` if an index is passed on that leads to a non-existing permission. Right now an empty permission is returned, maybe we should revert?

**Comment by alcueca** [comment]
@devtooligan, ready for review of audit fixes.

As discussed, instead of reverting on executing non existing permissions, I added a `check` function that highlights when a plan doesn't match the existing permissions.

File: contracts/utils/EmergencyBrake.sol [hunk 1]
```diff
@@ -1,191 +1,282 @@
...
+            // but that seems like a lesser evil versus allowing operators to override
+            // governance decisions.
             bytes4 signature_ = permission_.signature;
             require(
-                contact.hasRole(signature_, target),
+                host.hasRole(signature_, user),
                 "Permission not found"
             );
-            contact.revokeRole(signature_, target);
+            host.revokeRole(permission_.signature, user);
```
Comment (devtooligan) [inline contracts/utils/EmergencyBrake.sol:248]: @alcueca 
nit - is there a reason for creating the signature_ variable and then using it on the next line, but then not using it on the line after that?  seems more consistent either to use the variable in both places or just get rid of it altogether
```suggestion
            host.revokeRole(signature_, user);
```

**Review by devtooligan — APPROVED** [review]
audit fixes lgtm.  [the report has been updated](https://hackmd.io/@devtooligan/YieldEmergencyBrakeSecurityReview2022-12-201) with developer response and final reviewer response.

see one [nitpick comment](https://github.com/yieldprotocol/yield-utils-v2/pull/59/files#r1061076013)

**Pull request merged**